pilot/pkg/dns/dns_test.go

```diff
@@ -0,0 +1,311 @@
+	if initErr != nil {
+		t.Fatal(initErr)
+	}
+
+	testCases := []struct {
+		name                     string
+		host                     string
+		queryAAAA                bool
+		expected                 []dns.RR
+		expectResolutionFailure  bool
+		expectExternalResolution bool
+	}{
+		{
+			name:     "success: non k8s host in local cache",
+			host:     "www.google.com.",
+			expected: a("www.google.com.", []net.IP{net.ParseIP("1.1.1.1").To4()}),
+		},
+		{
+			name:     "success: non k8s host with search namespace yields cname",
+			host:     "www.google.com.ns1.svc.cluster.local.",
+			expected: cname("www.google.com.ns1.svc.cluster.local.", "www.google.com."),
+		},
+		{
+			name:                     "success: non k8s host not in local cache",
+			host:                     "www.bing.com.",
+			expectExternalResolution: true,
+		},
+		{
+			name:     "success: k8s host - fqdn",
+			host:     "productpage.ns1.svc.cluster.local.",
+			expected: a("productpage.ns1.svc.cluster.local.", []net.IP{net.ParseIP("9.9.9.9").To4()}),
+		},
+		{
+			name:     "success: k8s host - name.namespace",
+			host:     "productpage.ns1.",
+			expected: a("productpage.ns1.", []net.IP{net.ParseIP("9.9.9.9").To4()}),
+		},
+		{
+			name:     "success: k8s host - shortname",
+			host:     "productpage.",
+			expected: a("productpage.", []net.IP{net.ParseIP("9.9.9.9").To4()}),
+		},
+		{
+			name:     "success: k8s host (name.namespace) with search namespace yields cname",
+			host:     "productpage.ns1.ns1.svc.cluster.local.",
+			expected: cname("productpage.ns1.ns1.svc.cluster.local.", "productpage.ns1."),
+		},
+		{
+			name:     "success: k8s host - non local namespace - name.namespace",
+			host:     "reviews.ns2.",
+			expected: a("reviews.ns2.", []net.IP{net.ParseIP("10.10.10.10").To4()}),
+		},
+		{
+			name:     "success: k8s host - non local namespace - fqdn",
+			host:     "reviews.ns2.svc.cluster.local.",
+			expected: a("reviews.ns2.svc.cluster.local.", []net.IP{net.ParseIP("10.10.10.10").To4()}),
+		},
+		{
+			name:     "success: k8s host - non local namespace - name.namespace.svc",
+			host:     "reviews.ns2.svc.",
+			expected: a("reviews.ns2.svc.", []net.IP{net.ParseIP("10.10.10.10").To4()}),
+		},
+		{
+			name:                    "failure: k8s host - non local namespace - shortname",
+			host:                    "reviews.",
+			expectResolutionFailure: true,
+		},
+		{
+			name: "success: remote cluster k8s svc - same ns and different domain - fqdn",
+			host: "details.ns2.svc.cluster.remote.",
+			expected: a("details.ns2.svc.cluster.remote.",
+				[]net.IP{net.ParseIP("11.11.11.11").To4(), net.ParseIP("12.12.12.12").To4()}),
+		},
+		{
+			name:                    "failure: remote cluster k8s svc - same ns and different domain - name.namespace",
+			host:                    "details.ns2.",
+			expectResolutionFailure: true, // on home machines, the ISP may resolve to some generic webpage. So this test may fail on laptops
```

Comment:
any way we can not rely on external dns here?

Reply:
This is not.. on laptops in fios network, this ends up being resolved externally to soe dummy verizon catchall IP for all unknown DNS names (very annoying!). but on gke, it doesn't resolve at all.